feat: add UI trigger for query progress

Swiddis opened the pull request:
WIP -- for now I've added handling for non-terminal progress, I'm still working on how to use the trigger interface here: https://opensearch-project.github.io/OpenSearch-Dashboards/docs/#/../src/plugins/ui_actions/README. Never used something that requires an ID in a generic like this and getting syntax errors when I try to copy what other plugins are doing, so still WIP.

Issue: #25

Swiddis committed:
feat: Add handling for non-terminal progress

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
feat: Add spark job states to sql async interceptor

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
fix: Add uiActions dep

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
Merge remote-tracking branch 'upstream/main' into feature/loading-progress

Swiddis commented on public/search/sql_async_search_interceptor.ts:
```diff
@@ -77,18 +87,29 @@ export class SQLAsyncSearchInterceptor extends SearchInterceptor {
+            });
+            return false;
+          default:
+            if (request.params?.progress_query_id) {
+              this.uiActions.getTrigger(ASYNC_TRIGGER_ID).exec({
+                query_id: request.params.progress_query_id,
+                query_status: status,
+              });
+            }
```
Core logic is here: emit a UI action trigger for in-progress states. Rest of the PR is plumbing.

Question: How does a query pass in this request param? Is this part of the `search()` method call, and if so, is it as easy as `searchService.search({ progress_query_id: UUID4(), ... }, ...)`?

kavilla replied:
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/data/public/search/search_service.ts#L186

it will call this function which is passed in as a dependency to the search source.

discover creates a search source object. https://github.com/opensearch-project/OpenSearch-Dashboards/blob/5a4aba9014c04fbb1c4be17cdd40302ae0091126/src/plugins/discover/public/application/components/doc_views/context/utils/use_query_actions.ts#L33

kavilla commented on common/types.ts:
```diff
@@ -13,3 +14,25 @@ export interface FetchDataFrameContext {
 export type FetchFunction<T, P = void> = (params?: P) => Observable<T>;
+
+// ref: https://docs.aws.amazon.com/emr/latest/EMR-Serverless-UserGuide/job-states.html
+export enum SparkJobState {
```
nit: I think we can take out the ref to spark and the documentation link since these could be applied to things that aren't Spark related.

kavilla commented on public/plugin.tsx:
```diff
@@ -131,8 +138,7 @@ export class QueryEnhancementsPlugin
           searchBar: {
             showDatePicker: false,
             showFilterBar: false,
-            showDataSetsSelector: false,
```
might have been a merge conflict, but i think we need both these values.

kavilla commented on common/types.ts:
```diff
@@ -13,3 +14,25 @@ export interface FetchDataFrameContext {
+
+// ref: https://docs.aws.amazon.com/emr/latest/EMR-Serverless-UserGuide/job-states.html
+export enum SparkJobState {
+  SUBMITTED,
```
i think this needs to be set to a string like
```ts
SUBMITTED = 'SUBMITTED',
PENDING = 'PENDING',
...
```

Sauce: https://www.typescriptlang.org/docs/handbook/enums.html#string-enums

kavilla commented on public/search/sql_async_search_interceptor.ts:
```diff
@@ -77,18 +87,29 @@ export class SQLAsyncSearchInterceptor extends SearchInterceptor {
-        return false;
+      if (pollingResult) {
+        const statusStr: string = (pollingResult.body.meta.status as string).toUpperCase();
+        const status = SparkJobState[statusStr as keyof typeof SparkJobState];
```
from the above the string enum the string returned from statusstr can be compared in the switch

kavilla commented on public/search/sql_async_search_interceptor.ts:
```diff
@@ -77,18 +87,29 @@ export class SQLAsyncSearchInterceptor extends SearchInterceptor {
-        });
-        return false;
+      if (pollingResult) {
+        const statusStr: string = (pollingResult.body.meta.status as string).toUpperCase();
```
nit: can maybe put a helper function in utils that parses out the status? with optional chaining

kavilla commented on public/search/sql_async_search_interceptor.ts:
```diff
@@ -77,18 +87,29 @@ export class SQLAsyncSearchInterceptor extends SearchInterceptor {
+        const status = SparkJobState[statusStr as keyof typeof SparkJobState];
+        switch (status) {
+          case SparkJobState.SUCCESS:
+            return false;
```
unrelated to this PR but thought I get your feedback on this function.

when i refactored this a little I kept intact like the returning of a boolean but it returns false because the rxjs operator `takeWhile` needs to get the false value to stop polling. 

but it's confusing to read do you think it's worth to rename this function? or shift the logic so that in the utils function it negates the value here? Because return false when `SUCCESS` from onPollingSuccess could get confusing.

Swiddis replied:
It could be confusing at first, but I think it's generally derivable that the usage is to return `false` to halt polling and `true` to continue. There isn't anything inherently "keep polling"-like about `true` or `false`.

If we want to make it more clear I think the way to go would be to refactor the polling util to rely on a polling state enum, and return values like `Polling.DONE`, `Polling.CONTINUE`, `Polling.ERROR`. Could even make it more sophisticated with the specific statuses, having things like `TIMEOUT` or `ABORT`.

Swiddis replied:
Not sure if it's obvious by me introducing `JobState` to begin with, but I'm a big fan of using state machines and strong typing (i.e. lots of enums) to keep track of program state and handle errors.

kavilla commented on public/search/sql_async_search_interceptor.ts:
```diff
@@ -77,18 +87,29 @@ export class SQLAsyncSearchInterceptor extends SearchInterceptor {
+            });
+            return false;
+          default:
+            if (request.params?.progress_query_id) {
```
Apologies. I might be confused here and might need a little bit more context. 

Would you mind adding your approach details here: https://github.com/kavilla/queryEnhancements/issues/25? With how you intend this will be used in the rest of the application.

Once the request is here it is already in context of itself so if another request is fired I think it would have a different query ID and status. So here we are interceptor and won't return back a value to whoever invoked search until we want to return the dataframe value.

So it could maybe create the ID after this in here: https://github.com/kavilla/queryEnhancements/blob/main/public/search/sql_async_search_interceptor.ts#L114 and pass it to these functions and the ID wouldn't have to be passed in the request it just has to live temporarily when the request is intercepted (and emitted to anyone who is listening)

Swiddis replied:
The thing is that if we manage the request ID ourselves, we end up not able to communicate request progress to the outside -- the caller needs to know what request to listen on ahead of time, since we don't return any intermediary results via the existing search interface. The easiest way to accomplish this is by making the request query ID managed by the caller.

Swiddis replied:
Newest version unconditionally sends status with a generated ID, resolving

Swiddis committed:
fix dropped field from merge

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
update and document job state usage

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
fix lints

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
simplify parseJobState

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
Assert presence of job state

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
Merge remote-tracking branch 'upstream/main' into feature/loading-progress

Swiddis commented:
Keeping as draft while I figure out how the caller should consume the actions, might not need to filter by query ID param and might need to include additional info in the action context.

Swiddis committed:
Update interceptor to not require passed ID to send events

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
Finalize progress event sending

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis commented:
Alright: got something actually working. Sends query data to the `ASYNC_TRIGGER_ID` trigger. From the caller side you need to make a Subject that accepts the async query type. (To avoid cross-plugin imports this type also has to be redefined). Here's a sample method to create one of those (given a constant with the same async trigger id, and assuming the trigger ids have been registered):

```ts
import { Subject } from 'rxjs';
import { UiActionsStart, createAction } from '../../../ui_actions/public';
import { ASYNC_TRIGGER_ID, ASYNC_ACTION_ID } from '../../common';

interface AsyncQueryContext {
  queryId: string;
  queryStatus: string;
}

export function createQueryProgressSubject(uiActions: UiActionsStart): Subject<AsyncQueryContext> {
  const subject: Subject<AsyncQueryContext> = new Subject();
  uiActions.addTriggerAction(
    ASYNC_TRIGGER_ID,
    createAction<typeof ASYNC_ACTION_ID>({
      id: ASYNC_ACTION_ID,
      execute: async (context: unknown) => subject.next(context as AsyncQueryContext),
    })
  );
  return subject;
}
```

Then from the plugin `start` method you can create a subject:

```ts
const progressSubject = createQueryProgressAction(uiActions);
```

And pass this to the front-end where you can subscribe in order to update the progress:

```ts
const [progress, setProgress] = useState('NONE');
data.progressSubject.subscribe((v) => setProgress(v.queryStatus));
```

Pretty roundabout and far from ideal (requires lots of boilerplate to onboard the progress subject + query types + registration + passing all the refs around), but it gets the job done. We'll also require that the progress type is kept backwards-compatible, but otherwise should be pretty manageable.

ashwin-pc commented:
> Pretty roundabout and far from ideal (requires lots of boilerplate to onboard the progress subject + query types + registration + passing all the refs around)

Yeah for some of the simpler usecases we should add some abstractions inside the UI Actions plugin itself to make using it easier. Abstracting some this into the UI actions plugin directly would save future developers time here.

> We'll also require that the progress type is kept backwards-compatible

What do you mean by this?

Swiddis commented:
If we change the progress object type, other plugins will be depending on it, so it'll be a wider change -- by my count at least 3 plugins are going to be using this interface after it's ready (data/query workbench/observability)

This requirement probably exists already, but since we redefine the type in other plugins it won't be type-checked automatically if we do need to change it and track down all the downstream changes.

Swiddis committed:
Make state parsing robust against more types

The Flint API sometimes returns numeric HTTP statuses as well, for reasons not entirely understood to the commit author. Making the method more robust against any type simplifies using the interface, and checking for null results isn't any more annoying than it was before.

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis committed:
Fire initial submitted event

Signed-off-by: Simeon Widdis <sawiddis@amazon.com>

Swiddis commented:
https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7337
